### ISSUE TITLE
[Architecture] Define Build.Next

### ISSUE THREAD

**Issue by chmouel** [openshiftio/openshift.io#2488]
We need to think about build.next and what do we want to achieve with it.

* What is the current deficiency with the build system now.
*  What is the ideal build system
* How do we want to build it
* What are the depencies needed before.

This should be only a task for architecture and documentation.

**Comment by pradeepto** [comment]
Referencing https://github.com/openshiftio/openshift.io/issues/2488

**Comment by pradeepto** [comment]
Scribbled notes collated over a period of time: https://docs.google.com/document/d/1zpwR6rOdRWjQSHsbDYm_f_An2ND8lnJ599bo9AGsOm8/edit#

We need all the WHATs and WHYs before the HOWs.

**Comment by pradeepto** [comment]
Adding the slides and documents we got from Srikrishna today for more visibility.

- https://docs.google.com/presentation/d/1UYoNJKD_1DCeFqyT3Ozg-J71-zW8DJzUqFmQUQIJ5Xw/edit#slide=id.g3527c04894_0_136
- https://docs.google.com/presentation/d/1t7sFvvPG8BUwv9PdEoj0b93uXCOcb_sJ28ZrsAm22io/edit?ts=5aaa9e03#slide=id.g24cf0077fe_18_380
- https://docs.google.com/document/d/1MvfoIS3YvwIzAmCHGvQPhDRmmmrf7HP9QLa3EJ6aFf8/edit

**Comment by chmouel** [comment]
@krishnapaparaju My honest feedback, This slide here looks confusing to me : 

https://docs.google.com/presentation/d/1UYoNJKD_1DCeFqyT3Ozg-J71-zW8DJzUqFmQUQIJ5Xw/edit#slide=id.p

What are we trying to achieve ? As task executor for builds ? Is it to run concurrent tests or is it at merge time, is it to be able tu run tests on a collection of changesets  and make sure that all of them are tested independently ? 

This would redefine the way the developer works on github and do add a lot of complexities that only needed for *very* large projects, I am talking not your average large enteprise project, I am talking OpenStack / Linux Kernel kind of  size

Having said that if you really want complexities, this is something that openstack has been worked on for quite some time with a lot of thoughts, see this presentation at first  : 

https://blogs.rdoproject.org/2015/07/dive-into-zuul-gated-commit-system-2/

and this email from the main dev jim (who works at RH) : 

https://openstack.nimeyo.com/33063/openstack-dev-cross-repo-dependencies-in-zuul

but that's reallight tighten to Gerrit workflow,

Anyway if we really complexities (???) I would really advise to look at the current existing CI system before starting grand new concepts or trying to reproduce software patterns (task executor) into a CI/build pattern

Personally when I think about a build.next I think about a MVP that does mostly enough for us but yet fix a huge flaw in our current platform (not being multi tenant) which mean in a user story way :

- As a service delivery operator, I want to have a build system that simple and easy to understand but yet scalable to large project and I want this to be able to do multi tenancy.

And then building from there,

**Comment by lordofthejars** [comment]
Hello, first of all, I am going to express my opinion, so of course, you are free to think that I have no idea 😆 

I will focus Build.Next in a completely different way. Let me explain why. The architecture of having one central piece with queues in parallel, with some agents taking jobs and executing is exactly how Jenkins is implemented. And that's fine and it works but notices that it was done some time ago where our requirements are not the same as today. Why is Jenkins still here? Because it as something really good and it is that is versatile and adapts with its plugin system, but of course there is a payoff and all of us has suffered, complexity on updating plugins, incompatibilities, problems on updating configurations, strange behaviours because of some plugin, plugins that are not required but they are just there because they have been there for ages, and so on so forth.

Now let's stop one moment to think about what we need to do. Generally speaking (and I know I am missing a lot of things) we just want to do next thing:

* Compile a program
* Execute Tests
* Create a Docker container
* Push it to a registry
* Create a Kubernetes Deployment into Test env
* Create a Kubernetes Deployment into Stage/Prod env
* Switch from version 1 to version 2 (Canary Release, Blue/Green, ...)

You can see that 1) It is really simple 2) we know exactly the steps 3) it is clearly isolated 4) there is no downstreams/upstreams (in Microservices this concept should be avoided since every service must be auto deployable

So do we need a big architecture for such a simple thing? My opinion is no. What we need is something really simple because our workflow is simple, so let's think simple and things will become complex or not. So what I suggest is something that is been done in some new CI/CD products and it is that the container itself is able to run the CI/CD pipeline, so no master-slave or whatever.

Take as an example Escape (https://escape.ankyra.io/docs/what-is-escape/) which I would advocate using. It is two tools a CLI which builds everything and an Inventory (which is a service to store metadata and files), so the lifecycle is that no server is responsible to build something but the own project. For example, you do `escape build` and the project is built on where is running if it is your computer it is locally if it is a docker image is a docker image. Really simple right. Then this CLI when builds something it connects to inventory to store the generated files and the generated version. Then you can do `escape deploy -e test` which executes the script to deploy to test environment. And stores to inventory that artifact X version Y is at Test. And then you can do `escape deploy --from test to -e prod` and it takes the same artifact from inventory that is in test and executes the script but against `prod`. So it is simple, no master/slave nothing every project is responsible to build on their own.

The big advantage of that is that for updating you just need to update a CLI tool, that's all no servers, no shared configurations, nothing. Again simple fits our requirements and simplifies our life.

Of course, this is not enough, what we would need to develop is a Build service which accepts requests and spawns the container with project code and this CLI tool. Also should offer a way to prioritize builds, but it is a light and tight layer.

So I see the future of CI/CD with this micro builds approach where each project has the ability to execute itself the pipeline instead of the old way of server/agent approach.

That's my cents about the future of CI/CD. Have a good weekend 😋

**Comment by jaseemabid** [comment]
> So I see the future of CI/CD with this micro builds approach where each project has the ability to execute itself the pipeline instead of the old way of server/agent approach.

:+1: 

Master slave approach for CI/CD feels dated compared to what we can do with k8s and project specific containers.

**Comment by jaseemabid** [comment]
I'll try to chip in what I've heard from everyone else regarding Build.Next.
Since it's early days, please correct me on any of the points mentioned as we
are building foundations. I'll try to quote my sources as much as I can.

1. Jenkins is hugely popular in the space we are selling and we might have to
   support good old Jenkinsfile if customers ask us to. Even if we use something
   else under the hood for most users, there will be a Jenkins part of the
   puzzle we wont be able to throw away.

   Sources:

   - Brad mentioned the popularity in the townhall he did in Bangalore.
   - Adi mentioned the need to support customers with Jenkins if they ask us
     last week when a few of us met.

2. We are going GA with the current stack and we might not be able to abandon
   the flow users start with in the foreseeable future. We will support Jenkins
   builds for a while post GA.

3. Supporting all the interface exposed by Jenkins is an obvious nightmare, so
   we might come up with an alternate _simpler_ interface. One of the options is
   a yaml file with slightly limited capabilities.

   Designing the interface exposed by Next is probably going to be the bulk of
   the engineering work we might have to do. We need to support current setup,
   the business goals and not make the engineers go mad at the same time.

4. Scaling current setup is hard - Idler and Proxy are glorified band aids and
   I'm sure we all agree. The second half of Next is to run a sane service which
   we can

   1) Scale
   2) Monitor
   3) Understand
   4) Change as we wish

   I list all these, because there are substantial engineering problems
   associated with all of them.

** tldr; Realistically we are looking at a system which compiles yaml to
Jenkinsfile and run it for regular users and Jenkins itself for the power users
who would like to tweak it more and use something smarter than Idler to scale it
up **

Now the engineering reality of it. Almost none of the current code we have is
production ready. It works for a demo, but taking it any more seriously is a
mistake. Even if we go ahead with an architecture similar to what we have (I
think we will), this is my wish list/requirements as an engineer who was neck
deep in this for 4 months:

  1. Start with something that can be run on a developer's laptop. I know this
     is completely against the whole cloud native thing, but we are not there
     yet. It takes almost an hour to change a line of code and see it in action.
     It's a joke.

  2. Think much more seriously about testing, right now we have almost zero
     useful tests for all of build and we just hope that the production gods
     have mercy on us before every deploy. This severely limits the speed at
     which we can iterate and deploy. We manually test almost all changes by
     running a few quick starts.

  3. Build a service that we understand. The current team probably wrote less
     than 0.1% of the code that runs on production. Jenkins + 130 plugins might
     be something we can run on prod, but you need an army of support engineers
     for that.

  3. Throw away _all_ of f8 pipeline library. Its 88 groovy files copy pasted
     from each other with about 60% code duplication and no tests and docs.

  4. Be honest to what we sell and dog food our build story.

**Comment by kpiwko** [comment]
The fact of master - slave approach being dated fully resonates with me.

I believe that we would get further by just having Ansible and S2I in the mix instead of Jenkins Master.


> I'll try to chip in what I've heard from everyone else regarding Build.Next.
Since it's early days, please correct me on any of the points mentioned as we
are building foundations. I'll try to quote my sources as much as I can.
>
> Jenkins is hugely popular in the space we are selling and we might have to
support good old Jenkinsfile if customers ask us to. Even if we use something
else under the hood for most users, there will be a Jenkins part of the
puzzle we wont be able to throw away.

This we could solve by having Jenkins only for users that are paying for the infra - no powerusers capabilities in free tier. And hence no Idler.

> Sources:
> Brad mentioned the popularity in the townhall he did in Bangalore.
Adi mentioned the need to support customers with Jenkins if they ask us
last week when a few of us met.
>We are going GA with the current stack and we might not be able to abandon
the flow users start with in the foreseeable future. We will support Jenkins
builds for a while post GA.

I believe that we need to write down what this GA really means. My understanding is that we don't commit to stability of API yet - we are just announcing **G**eneral **A**vailability of the service.

> Supporting all the interface exposed by Jenkins is an obvious nightmare, so
we might come up with an alternate simpler interface. One of the options is
a yaml file with slightly limited capabilities.

Does anybody know what it Travis CI using to run travis.yaml descriptions?

> Designing the interface exposed by Next is probably going to be the bulk of
the engineering work we might have to do. We need to support current setup,
the business goals and not make the engineers go mad at the same time.

> Scaling current setup is hard - Idler and Proxy are glorified band aids and
I'm sure we all agree. The second half of Next is to run a sane service which
we can
>
> Scale
> Monitor
> Understand
> Change as we wish
> I  list all these, because there are substantial engineering problems
associated with all of them.

Could we have a serverless approach for the build service instead? My understanding is that s2i would pretty much do that.

> tldr; Realistically we are looking at a system which compiles yaml to
Jenkinsfile and run it for regular users and Jenkins itself for the power users
who would like to tweak it more and use something smarter than Idler to scale it
up
>
>Now the engineering reality of it. Almost none of the current code we have is
production ready. It works for a demo, but taking it any more seriously is a
mistake. Even if we go ahead with an architecture similar to what we have (I
think we will), this is my wish list/requirements as an engineer who was neck
deep in this for 4 months:

> Start with something that can be run on a developer's laptop. I know this
is completely against the whole cloud native thing, but we are not there
yet. It takes almost an hour to change a line of code and see it in action.
It's a joke.

> Think much more seriously about testing, right now we have almost zero
useful tests for all of build and we just hope that the production gods
have mercy on us before every deploy. This severely limits the speed at
which we can iterate and deploy. We manually test almost all changes by
running a few quick starts.

For that the step mentioned above would help - as making service available at laptop would mean it is more decoupled and easier to test using lower levels than e2e tests.

> Build a service that we understand. The current team probably wrote less
than 0.1% of the code that runs on production. Jenkins + 130 plugins might
be something we can run on prod, but you need an army of support engineers
for that.

> Throw away all of f8 pipeline library. Its 88 groovy files copy pasted
from each other with about 60% code duplication and no tests and docs.

Could we have a common functionality available as Ansible playbooks?

> Be honest to what we sell and dog food our build story.

+1, we should use that system to build openshift.io itself to prove it is the right approach.

**Comment by jaseemabid** [comment]
> Does anybody know what it Travis CI using to run travis.yaml descriptions?

https://github.com/travis-ci/travis-build

Regarding configuration with yaml, I love what https://drone.io did. Its a superset of the docker-compose format and their examples look really neat.

**Comment by chmouel** [comment]
> Does anybody know what it Travis CI using to run travis.yaml descriptions

https://github.com/travis-ci/travis-ci <https://github.com/travis-ci/travis-ci>

**Comment by bmicklea** [comment]
Just to make sure the point is captured here as well as in email threads. Customers need to have some way to use Jenkins, but that doesn't mean that if we create a build service that we have to be able to migrate their Jenkins files to it. We could give them several options:
- Add Jenkins to their tenant space so they're "spending" the resources to use it
- Document how they can use OSIO with their own Jenkins (assuming it's accessible to OSIO)
- Offer a centralized build service that they can use for free

There may be other options too. @maxandersen please add/clarify.
